### PR TITLE
Extract from internal salesforce lib

### PR DESCRIPTION
**Pull request by iky**

### CODE REVIEW

File: nameko_salesforce/streaming/client.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        pipe.execute()
+
+
+class MessageHandler(BayeuxMessageHandler):
+
+    client = StreamingClient()
+
+    def handle_message(self, message):
+        args, kwargs = self.get_worker_args(message)
+        context_data = {}
```
Comment (iky) [inline nameko_salesforce/streaming/client.py]: Good point. Besides the data, there is `client ID`, `replay ID` and the notification creation timestamp.

Maybe the `client` and `replay` IDs we could dump to the context?

`Client ID` is unique per client instance connected to the server
`Replay ID` is unique per notification message in a channel

File: tests/api/test_client.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+
+@pytest.fixture(autouse=True)
+def mock_salesforce_login():
+    with patch('simple_salesforce.api.SalesforceLogin') as SalesforceLogin:
+        SalesforceLogin.return_value = 'session_id', 'abc.salesforce.com'
+        yield
+
+
+@pytest.fixture
```
Comment (iky) [inline tests/api/test_client.py]: True, added

File: tests/api/test_dependency.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        expected_error = (
+            '`{}` configuration does not contain mandatory `{}` key'
+            .format(constants.CONFIG_KEY, key))
+        assert str(exc.value) == expected_error
+
+    def test_get_dependency(self, config, make_salesforce_api_provider):
+        salesforce_api_provider = make_salesforce_api_provider()
+        salesforce_api_provider.setup()
+        worker_ctx = {}
```
Comment (iky) [inline tests/api/test_dependency.py]: Very true, replaced with Mock

File: nameko_salesforce/api/push_topics.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        notify_for_operation_update=True,
+        notify_for_operation_delete=True,
+        notify_for_operation_undelete=True,
+    ):
+        """
+        Update or create Push Topic object
+
+        Note that the update or create operation is not atomic and there
+        is a race condition with multiple services trying to declare the topic
+        at the same time.
```
Comment (mattbennett) [inline nameko_salesforce/api/push_topics.py]: OK fine. If there's no mitigation, what is the fallout when who services do try to create at the same time? Does the losing service raise and terminate?

File: nameko_salesforce/streaming/client.py (hunk 1 above)
Comment (mattbennett) [inline nameko_salesforce/streaming/client.py]: Yeah I think both of these would be helpful 👍

Comment (iky) [inline nameko_salesforce/streaming/client.py]: :+1:

File: tests/streamig/conftest.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import pytest
```
Comment (iky) [inline tests/streamig/conftest.py]: :+1: fixed

File: tests/streamig/test_end_to_end.py [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+    return Mock()
+
+
+@pytest.fixture
+def waiter():
+    return Event()
+
+
+@pytest.fixture
+def make_salesforce_server(
```
Comment (iky) [inline tests/streamig/test_end_to_end.py]: It should keep the "make" verb in it as it returns a function. I can add the "stub" at the end, but not sure how that improves the already quite long name?

**Comment by mattbennett** [comment]
:+1: Great stuff

**Pull request merged**